File: backend/app/schemas/financial.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.models import ChargeType, DocumentType, PaymentMethod, PaymentStatus


class ChargeCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: ChargeType
    description: str = Field(min_length=3, max_length=250)
    amount: Decimal = Field(gt=0, decimal_places=2)


class ChargeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    rental_id: uuid.UUID
    type: ChargeType
    description: str
    amount: Decimal
    created_at: datetime


class PaymentCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    method: PaymentMethod
    amount: Decimal = Field(gt=0, decimal_places=2)
    paid_at: datetime | None = None
    reference: str | None = Field(default=None, max_length=120)


class PaymentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    rental_id: uuid.UUID
    method: PaymentMethod
    status: PaymentStatus
    amount: Decimal
    paid_at: datetime
    reference: str | None
    created_at: datetime


class FinancialSummary(BaseModel):
    rental_id: uuid.UUID
    charges: list[ChargeOut]
    payments: list[PaymentOut]
    charge_total: Decimal
    paid_total: Decimal
    balance_due: Decimal


class DocumentCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: DocumentType


class DocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    rental_id: uuid.UUID
    type: DocumentType
    version: int
    content_sha256: str
    created_at: datetime
    download_url: str = ""

File: frontend/src/api/financial.ts
import { apiDownload, apiFetch } from '@/api/client'

export type ChargeType = 'DISCOUNT' | 'CLEANING' | 'DAMAGE' | 'ADJUSTMENT'
export type PaymentMethod = 'CASH' | 'PIX' | 'CARD' | 'TRANSFER' | 'OTHER'
export type DocumentType = 'CONTRACT' | 'PICKUP_TERM' | 'RETURN_TERM' | 'RECEIPT' | 'EXTRA_RECEIPT'

export interface ChargeData { id: string; type: string; description: string; amount: string; created_at: string }
export interface PaymentData { id: string; method: PaymentMethod; status: string; amount: string; paid_at: string; reference: string | null }
export interface FinancialSummary { rental_id: string; charges: ChargeData[]; payments: PaymentData[]; charge_total: string; paid_total: string; balance_due: string }
export interface DocumentData { id: string; type: DocumentType; version: number; content_sha256: string; created_at: string; download_url: string }

export function getFinancial(rentalId: string): Promise<FinancialSummary> { return apiFetch(`/rentals/${rentalId}/financial`) }
export function addCharge(rentalId: string, payload: { type: ChargeType; description: string; amount: number }, key: string): Promise<ChargeData> {
  return apiFetch(`/rentals/${rentalId}/charges`, { method: 'POST', headers: { 'Idempotency-Key': key }, body: JSON.stringify(payload) })
}
export function addPayment(rentalId: string, payload: { method: PaymentMethod; amount: number; reference?: string | null }, key: string): Promise<PaymentData> {
  return apiFetch(`/rentals/${rentalId}/payments`, { method: 'POST', headers: { 'Idempotency-Key': key }, body: JSON.stringify(payload) })
}
export function listDocuments(rentalId: string): Promise<DocumentData[]> { return apiFetch(`/rentals/${rentalId}/documents`) }
export function generateDocument(rentalId: string, type: DocumentType, key: string): Promise<DocumentData> {
  return apiFetch(`/rentals/${rentalId}/documents`, { method: 'POST', headers: { 'Idempotency-Key': key }, body: JSON.stringify({ type }) })
}
export async function downloadDocument(document: DocumentData): Promise<void> {
  const blob = await apiDownload(document.download_url.replace('/api/v1', ''))
  const url = URL.createObjectURL(blob)
  const link = window.document.createElement('a')
  link.href = url
  link.download = `${document.type.toLowerCase()}-v${document.version}.pdf`
  link.click()
  URL.revokeObjectURL(url)
}
